core/src/main/java/ai/timefold/solver/core/impl/domain/variable/ExternalizedListVariableStateSupply.java

```diff
@@ -4,16 +4,16 @@
 import java.util.Map;
 import java.util.Objects;
 
+import ai.timefold.solver.core.api.domain.metamodel.ElementLocation;
+import ai.timefold.solver.core.api.domain.metamodel.LocationInList;
 import ai.timefold.solver.core.api.score.director.ScoreDirector;
 import ai.timefold.solver.core.impl.domain.variable.descriptor.ListVariableDescriptor;
-import ai.timefold.solver.core.impl.heuristic.selector.list.ElementLocation;
-import ai.timefold.solver.core.impl.heuristic.selector.list.LocationInList;
 
 final class ExternalizedListVariableStateSupply<Solution_>
```

Comment:
Here is one of the places we could enhance the experience by modifying the default contract so that `ExternalizedListVariableStateSupply::getLocationInList` returns `LocationInList`. I'm uncertain whether calling `ensureAssigned()` would disrupt existing code, but if not, it would be logical to apply this validation when retrieving an element
`getLocationInList`.

My point is to return something that can be used without casting or converting with `ensureAssigned`. On the other hand, I understand that changing the contract could impact many other areas.

Reply:
This change to require casting is very deliberate and it is intended to _prevent user error_ - it is a way to prevent NPEs. If you know something is supposed to be assigned, you better say it out loud in the code, and make sure that you handle the error case.

You'll see this pattern much more in new Java code. With sealed classes and the advent of pattern matching, this is going to be the new way of writing Java code. If you haven't yet, watch this talk to understand what is already possible in Java, what is soon going to be possible, and how that's going to make Java code better and less buggy:

https://www.youtube.com/watch?v=LgquOtnjMJ8